language/vm/vm_runtime/vm_runtime_types/src/value.rs

```diff
@@ -184,228 +126,471 @@ pub struct RootAccessPath {
 
 /// A GlobalRef holds the reference to the data and a shared reference to the root so
 /// status flags and reference count can be properly managed
-#[derive(Debug, Clone)]
+#[derive(PartialEq, Eq, Debug, Clone)]
 pub struct GlobalRef {
     root: Rc<RefCell<RootAccessPath>>,
     reference: MutVal,
 }
 
-impl Clone for MutVal {
-    fn clone(&self) -> Self {
-        MutVal(Rc::new(RefCell::new(self.peek().clone())))
+// All implementation here is private to this module and the effective logic in
+// working with values in the VM.
+impl ValueImpl {
+    fn into_value(self) -> VMResult<Value> {
+        match self {
+            ValueImpl::Invalid => Err(VMStatus::new(StatusCode::INTERNAL_TYPE_ERROR)),
```

Comment:
A comment that this should never happen because of the bytecode verifier might be useful?

Reply:
as we talked about the comment in the error file should be enough.
Having said so we want to evolve this with invariants/asserts as much as possible (MIRAI being an option) so I'd leave it as is for now and do something later on